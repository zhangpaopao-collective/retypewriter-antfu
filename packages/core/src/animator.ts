import { calculatePatches, diff } from './patch'
import type { Patch } from './types'

export interface AnimatorStep {
  index: number
  char?: string
  cursor: number
  content: string
}

export function* createAnimator(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let _input = input
  for (let patchIndex = 0; patchIndex < patches.length; patchIndex++) {
    const patch = patches[patchIndex]
    if (patch.type === 'insert') {
      for (let i = 0; i < patch.text.length; i++) {
        const index = patch.from + i
        _input = _input.slice(0, index) + patch.text[i] + _input.slice(index)
        yield {
          index: patchIndex,
          char: patch.text[i],
          cursor: index,
          content: _input,
        }
      }
    }
    else {
      for (let i = 0; i < patch.length; i++) {
        const index = patch.from - i
        _input = _input.slice(0, index - 1) + _input.slice(index)
        yield {
          index: patchIndex,
          cursor: index,
          content: _input,
        }
      }
    }
  }
}

export function simpleAnimator(input: string, output: string) {
  const delta = diff(input, output)
  const patches = calculatePatches(delta)

  return createAnimator(input, patches)
}
