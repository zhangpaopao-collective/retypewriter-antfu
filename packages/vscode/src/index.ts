import { commands, window } from 'vscode'

export interface Snapshot {
  content: string
  time: number
}

export function activate() {
  window.showInformationMessage('Hi')

  const snapMap = new Map<string, Snapshot[]>()

  commands.registerCommand('retypewriter.snap', () => {
    const doc = window.activeTextEditor?.document
    if (!doc)
      return
    const path = doc.uri.fsPath
    if (!snapMap.has(path))
      snapMap.set(path, [])
    snapMap.get(path)!.push({
      content: doc.getText(),
      time: Date.now(),
    })
  })

  commands.registerCommand('retypewriter.play', () => {
    const doc = window.activeTextEditor?.document
    if (!doc)
      return
    const path = doc.uri.fsPath
    const snaps = snapMap.get(path)
    if (!snaps?.length)
      window.showErrorMessage('No snapshots found')
      // return

    // const lastSnap = snaps[snaps.length - 1]
  })
}

export function deactivate() {

}
