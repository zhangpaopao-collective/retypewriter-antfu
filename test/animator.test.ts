import { describe, expect, it } from 'vitest'
import { input, output } from '../examples'
import { calculatePatches, createAnimator, diff } from '../src'

describe('animator', () => {
  it('animator', () => {
    const delta = diff(input, output)
    const patches = calculatePatches(delta)

    const animator = createAnimator(input, patches)

    expect([...animator]).toMatchSnapshot()
  })
})
