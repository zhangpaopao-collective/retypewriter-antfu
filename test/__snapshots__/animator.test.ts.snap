// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`animator > animator 1`] = `
[
  {
    "cursor": 35,
    "output": "
  patches.push({
    type: 'inser',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 34,
    "output": "
  patches.push({
    type: 'inse',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 33,
    "output": "
  patches.push({
    type: 'ins',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 32,
    "output": "
  patches.push({
    type: 'in',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 31,
    "output": "
  patches.push({
    type: 'i',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 30,
    "output": "
  patches.push({
    type: '',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 29,
    "output": "
  patches.push({
    type: 'r',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 30,
    "output": "
  patches.push({
    type: 're',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 31,
    "output": "
  patches.push({
    type: 'rem',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 32,
    "output": "
  patches.push({
    type: 'remo',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 33,
    "output": "
  patches.push({
    type: 'remov',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 34,
    "output": "
  patches.push({
    type: 'remova',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 35,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    text: change[1],
  })
",
  },
  {
    "cursor": 64,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    tex: change[1],
  })
",
  },
  {
    "cursor": 63,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    te: change[1],
  })
",
  },
  {
    "cursor": 62,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    t: change[1],
  })
",
  },
  {
    "cursor": 61,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    : change[1],
  })
",
  },
  {
    "cursor": 60,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    l: change[1],
  })
",
  },
  {
    "cursor": 61,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    le: change[1],
  })
",
  },
  {
    "cursor": 62,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    len: change[1],
  })
",
  },
  {
    "cursor": 63,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    leng: change[1],
  })
",
  },
  {
    "cursor": 64,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    lengt: change[1],
  })
",
  },
  {
    "cursor": 65,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1],
  })
",
  },
  {
    "cursor": 77,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].,
  })
",
  },
  {
    "cursor": 78,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].l,
  })
",
  },
  {
    "cursor": 79,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].le,
  })
",
  },
  {
    "cursor": 80,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].len,
  })
",
  },
  {
    "cursor": 81,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].leng,
  })
",
  },
  {
    "cursor": 82,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].lengt,
  })
",
  },
  {
    "cursor": 83,
    "output": "
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].length,
  })
",
  },
]
`;
